src/Init/System/Mutex.lean

```diff
@@ -0,0 +1,123 @@
+-/
+def BaseMutex : Type := BaseMutexImpl.type
+
+instance : Nonempty BaseMutex := BaseMutexImpl.property
+
+/-- Creates a new `BaseMutex`. -/
+@[extern "lean_io_basemutex_new"]
+opaque BaseMutex.new : BaseIO BaseMutex
+
+/--
+Locks a `BaseMutex`.  Waits until no other thread has locked the mutex.
+
+The current thread must not have already locked the mutex.
+Reentrant locking is undefined behavior (inherited from the C++ implementation).
+-/
+@[extern "lean_io_basemutex_lock"]
+opaque BaseMutex.lock (mutex : @& BaseMutex) : BaseIO Unit
+
+/--
+Unlocks a `BaseMutex`.
+
+The current thread must have already locked the mutex.
+Unlocking an unlocked mutex is undefined behavior (inherited from the C++ implementation).
+-/
+@[extern "lean_io_basemutex_unlock"]
+opaque BaseMutex.unlock (mutex : @& BaseMutex) : BaseIO Unit
+
+private opaque CondvarImpl : NonemptyType.{0}
+
+/-- Condition variable. -/
+def Condvar : Type := CondvarImpl.type
+
+instance : Nonempty Condvar := CondvarImpl.property
+
+/-- Creates a new condition variable. -/
+@[extern "lean_io_condvar_new"]
+opaque Condvar.new : BaseIO Condvar
+
+/-- Waits until another thread calls `notifyOne` or `notifyAll`. -/
+@[extern "lean_io_condvar_wait"]
+opaque Condvar.wait (condvar : @& Condvar) (mutex : @& BaseMutex) : BaseIO Unit
+
+/-- Wakes up a single other thread executing `wait`. -/
+@[extern "lean_io_condvar_notify_one"]
+opaque Condvar.notifyOne (condvar : @& Condvar) : BaseIO Unit
+
+/-- Wakes up all other threads executing `wait`. -/
+@[extern "lean_io_condvar_notify_all"]
+opaque Condvar.notifyAll (condvar : @& Condvar) : BaseIO Unit
+
+/-- Waits on the condition variable until the predicate is true. -/
+def Condvar.waitUntil [Monad M] [MonadLift BaseIO M]
+    (condvar : Condvar) (mutex : BaseMutex) (pred : M Bool) : M Unit := do
+  while !(← pred) do
+    condvar.wait mutex
+
+/--
+Mutual exclusion primitive (lock) guarding shared state of type `α`.
+
+The type `Mutex α` is similar to `IO.Ref α`,
+except that concurrent accesses are guarded by a mutex
+instead of atomic pointer operations and busy-waiting.
+-/
+structure Mutex (α : Type) where private mk ::
+  private ref : IO.Ref α
+  mutex : BaseMutex
+
+instance [Nonempty α] : Nonempty (Mutex α) :=
+  let ⟨ref⟩ := inferInstanceAs (Nonempty _)
+  let ⟨mutex⟩ := inferInstanceAs (Nonempty _)
+  ⟨{ref, mutex}⟩
+
+instance : Coe (Mutex α) BaseMutex where coe := Mutex.mutex
+
+/-- Creates a new mutex. -/
+def Mutex.new (a : α) : BaseIO (Mutex α) :=
+  return { ref := ← mkRef a, mutex := ← BaseMutex.new }
+
+/--
+`AtomicT α M` is the monad that can be atomically executed inside a `Mutex α`,
+with outside monad `M`.
+The action has access to the state `α` of the mutex (via `get` and `set`).
+-/
+abbrev AtomicT := StateRefT' IO.RealWorld
+
+/-- `mutex.atomically k` runs `k` with access to the mutex's state while locking the mutex. -/
+def Mutex.atomically [Monad M] [MonadLiftT BaseIO M] [MonadFinally M]
```

Comment:
Please let's keep the current naming convention for now.

Reply:
I think I wrote this a couple weeks back when `Monad M` was just gaining traction.  It's uppercase because it's a type.  I've changed it to `m` in this PR.  But there's plenty of other occurrences, check out `rg 'Monad M'`.

Reply:
Yes, it seems they all come from @EdAyers :)